### PR TITLE
Spatter augmentation tensor implementation

### PR DESCRIPTION
**Pull request by r-abishek**
@rrawther @asalmanp This PR adds Spatter augmentation to RPP.

Adds host and hip implementations for spatter augmentation.
Adds the corresponding test suite on host and hip for running both unit and performance tests.

### CODE REVIEW

File: src/include/cpu/rpp_cpu_common.hpp
Comment (rrawther) [inline src/include/cpu/rpp_cpu_common.hpp:2084]: please check if pSpatterMaskInv is same for all

File: src/include/cpu/rpp_cpu_simd.hpp
Comment (rrawther) [inline src/include/cpu/rpp_cpu_simd.hpp:798]: keep these masks as constant globals since it is repeatedly used

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (rrawther) [inline src/modules/cpu/host_tensor_augmentations.hpp:9272]: What is the difference between the spatterMask and InvSpatterMask

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (rrawther) [inline src/modules/cpu/host_tensor_augmentations.hpp:9264]: Does this work for image sizes more than 1920x1080

File: src/modules/hip/kernel/spatter.hpp
Comment (rrawther) [inline src/modules/hip/kernel/spatter.hpp:6]: can't you avoid different definitions using Template argument for the srcPtr?

**Review by rrawther — CHANGES_REQUESTED** [review]
please address review comments

File: src/include/cpu/rpp_cpu_common.hpp
Comment (r-abishek) [reply src/include/cpu/rpp_cpu_common.hpp:2084]: Yes, the spatterMask/spatterMaskInv only contains the mask for the spattered regions in the image. So they should be the same for the R/G/B channels.

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (r-abishek) [reply src/modules/cpu/host_tensor_augmentations.hpp:9272]: The spatterMask is basically 1 - spatterMaskInv and both are needed. Instead of computing that at runtime, we have both statically initialized.

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (r-abishek) [reply src/modules/cpu/host_tensor_augmentations.hpp:9264]: Actually no. Had planned to have the max as 4K (3840 x 2160), but the static declaration becomes too large, so decided to cap it at 1920 x 1080. I have mentioned it in the header documentation -> https://github.com/GPUOpen-ProfessionalCompute-Libraries/rpp/pull/87/files/ebc88e3851dc6ea7c3b723eb3659667717b61c1a#diff-72f99bf4cde5647f79cb11784b92bfa86370aedb5fb7ebc9a8b114cd4405c988R237

File: src/modules/hip/kernel/spatter.hpp
Comment (r-abishek) [reply src/modules/hip/kernel/spatter.hpp:6]: Templating could be used if every arithmetic inside the 4 functions are exactly the same except for the data-type, which can be templated as <T>. But in this case, we have different computations for each function.

File: src/include/cpu/rpp_cpu_simd.hpp
Comment (r-abishek) [reply src/include/cpu/rpp_cpu_simd.hpp]: Done

File: src/modules/hip/kernel/spatter.hpp
Comment (asalmanp) [inline src/modules/hip/kernel/spatter.hpp:27]: what's this 0.0039216 magic number? can you add a comment for every const number used in these kernels?

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/modules/hip/kernel/spatter.hpp
Comment (asalmanp) [inline src/modules/hip/kernel/spatter.hpp:238]: you should not allocate memory here during the launch, the allocation should occur outside of this function and pass a pointer to this hip_exec_spatter_tensor. Please remove line 238-241 from here.

File: src/modules/hip/kernel/spatter.hpp
Comment (asalmanp) [inline src/modules/hip/kernel/spatter.hpp:318]: same comment as above. freeing a memory here is not correct and safe. freeing memory for the masks should be outside of this function.

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/modules/rppt_tensor_augmentations.cpp
Comment (asalmanp) [inline src/modules/rppt_tensor_augmentations.cpp:1223]: same comment as above. please remove allocation from this function.

File: src/modules/rppt_tensor_augmentations.cpp
Comment (asalmanp) [inline src/modules/rppt_tensor_augmentations.cpp:1274]: move hipFree outside of this function.

File: utilities/rpp-performancetests/HIP_NEW/Tensor_hip_pkd3.cpp
Comment (asalmanp) [inline utilities/rpp-performancetests/HIP_NEW/Tensor_hip_pkd3.cpp:1056]: we talked about these comments before that you would need to pass these as parameters to the test. When are you going to make these changes?

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: src/modules/hip/kernel/spatter.hpp
Comment (r-abishek) [reply src/modules/hip/kernel/spatter.hpp]: Done. #define is now used to initialize a common 1/255

File: src/modules/hip/kernel/spatter.hpp
Comment (r-abishek) [reply src/modules/hip/kernel/spatter.hpp]: I've moved the allocation out. Now pre-allocating temp hip buffers like for other functions.
I presume copying new data into a pre-allocated buffer is okay to have in the call path?

File: src/modules/hip/kernel/spatter.hpp
Comment (r-abishek) [reply src/modules/hip/kernel/spatter.hpp]: Done

File: src/modules/rppt_tensor_augmentations.cpp
Comment (r-abishek) [reply src/modules/rppt_tensor_augmentations.cpp]: Done

File: src/modules/rppt_tensor_augmentations.cpp
Comment (r-abishek) [reply src/modules/rppt_tensor_augmentations.cpp]: Done

File: utilities/rpp-performancetests/HIP_NEW/Tensor_hip_pkd3.cpp
Comment (r-abishek) [reply utilities/rpp-performancetests/HIP_NEW/Tensor_hip_pkd3.cpp:1056]: Since every function in the test suite has a different number and type of parameters taken from the user, this change might not actually be a direct one. Some functions take a single u8/uint/f32 or some functions take whole arrays passed in, and others have tensor based ROIs. Will take a look on this in a separate test suite PR. Should be after the next couple of essential PRs ricap/resize/colortwist/cmn are merged.

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (rrawther) [inline src/modules/cpu/host_tensor_augmentations.hpp:9264]: Is SPATTER_MAX_WIDTH and height user parameters?

File: src/modules/hip/handlehip.cpp
Comment (rrawther) [inline src/modules/hip/handlehip.cpp:250]: it is better not name the array with .floatmem (is this a union?). Please keep it simple.

**Review by rrawther — APPROVED** [review]
Can be merged after @asalmanp approves HIP changes

File: src/modules/cpu/host_tensor_augmentations.hpp
Comment (r-abishek) [reply src/modules/cpu/host_tensor_augmentations.hpp:9264]: No, the SPATTER_MAX_WIDTH is defined at compile time. 1920 for now.

File: src/modules/hip/handlehip.cpp
Comment (r-abishek) [reply src/modules/hip/handlehip.cpp:250]: This one actually was developed previously for rpphandle from batchpd. Just used it here. Shall look into if changes can be made in a subsequent pr.

File: src/modules/hip/kernel/spatter.hpp
Comment (r-abishek) [reply src/modules/hip/kernel/spatter.hpp:244]: It's defined in src/include/func_specific/spatter_mask.hpp

**Review by asalmanp — APPROVED** [review]

**Pull request merged**
